panic in net.forEachRoute when starting container via proxy

````
$ eval $(weave env)
$ docker run --name=foo -ti ubuntu
root@foo:/# exit
exit
$ docker run --name=foo1 --hostname=foo.weave.local -ti ubuntu
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x20 pc=0x458102]

goroutine 1 [running]:
github.com/weaveworks/weave/net.forEachRoute(0xc8200a00f0, 0xc820039be8, 0x0, 0x0)
	/go/src/github.com/weaveworks/weave/net/route.go:49 +0x232
github.com/weaveworks/weave/net.CheckRouteExists(0x55c068, 0x5, 0xc820039c68, 0x10, 0x10, 0x0)
	/go/src/github.com/weaveworks/weave/net/route.go:63 +0xa8
github.com/weaveworks/weave/net.EnsureInterfaceAndMcastRoute(0x55c068, 0x5, 0x0, 0x0, 0x0)
	/go/src/github.com/weaveworks/weave/net/if.go:56 +0x338
main.checkNetwork(0x0, 0x0)
	/go/src/github.com/weaveworks/weave/prog/weavewait/check_network_iface_mcast.go:10 +0x38
main.main()
	/go/src/github.com/weaveworks/weave/prog/weavewait/main.go:20 +0x6b

goroutine 6 [chan send]:
github.com/weaveworks/weave/vendor/github.com/vishvananda/netlink.LinkSubscribe.func2(0xc820014120, 0xc82000a0e0)
	/go/src/github.com/weaveworks/weave/vendor/github.com/vishvananda/netlink/link_linux.go:847 +0x2e6
created by github.com/weaveworks/weave/vendor/github.com/vishvananda/netlink.LinkSubscribe
	/go/src/github.com/weaveworks/weave/vendor/github.com/vishvananda/netlink/link_linux.go:850 +0x107

goroutine 17 [runnable]:
syscall.Syscall(0x3, 0x4, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/syscall/asm_linux_amd64.s:18 +0x5
syscall.Close(0x4, 0x0, 0x0)
	/usr/local/go/src/syscall/zsyscall_linux_amd64.go:296 +0x4a
github.com/weaveworks/weave/vendor/github.com/vishvananda/netlink/nl.(*NetlinkSocket).Close(0xc82009e220)
	/go/src/github.com/weaveworks/weave/vendor/github.com/vishvananda/netlink/nl/nl_linux.go:325 +0x24
github.com/weaveworks/weave/vendor/github.com/vishvananda/netlink.RouteSubscribe.func1(0xc8200b6060, 0xc82009e220)
	/go/src/github.com/weaveworks/weave/vendor/github.com/vishvananda/netlink/route_linux.go:306 +0x44
created by github.com/weaveworks/weave/vendor/github.com/vishvananda/netlink.RouteSubscribe
	/go/src/github.com/weaveworks/weave/vendor/github.com/vishvananda/netlink/route_linux.go:307 +0xe6

goroutine 18 [runnable]:
syscall.Syscall6(0x2d, 0x4, 0xc8200ca000, 0x1000, 0x0, 0xc82001ec48, 0xc82001ec44, 0x6c, 0x1000, 0x0)
	/usr/local/go/src/syscall/asm_linux_amd64.s:44 +0x5
syscall.recvfrom(0x4, 0xc8200ca000, 0x1000, 0x1000, 0x0, 0xc82001ec48, 0xc82001ec44, 0x40e819, 0x0, 0x0)
	/usr/local/go/src/syscall/zsyscall_linux_amd64.go:1712 +0x91
syscall.Recvfrom(0x4, 0xc8200ca000, 0x1000, 0x1000, 0x0, 0x1000, 0x0, 0x0, 0x0, 0x0)
	/usr/local/go/src/syscall/syscall_unix.go:244 +0xb8
github.com/weaveworks/weave/vendor/github.com/vishvananda/netlink/nl.(*NetlinkSocket).Receive(0xc82009e220, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/weaveworks/weave/vendor/github.com/vishvananda/netlink/nl/nl_linux.go:341 +0xae
github.com/weaveworks/weave/vendor/github.com/vishvananda/netlink.RouteSubscribe.func2(0xc8200b6000, 0xc82009e220)
	/go/src/github.com/weaveworks/weave/vendor/github.com/vishvananda/netlink/route_linux.go:312 +0x61
created by github.com/weaveworks/weave/vendor/github.com/vishvananda/netlink.RouteSubscribe
	/go/src/github.com/weaveworks/weave/vendor/github.com/vishvananda/netlink/route_linux.go:324 +0x115
Error response from daemon: Failure during network configuration for container 08f8447ca51cf71a1c6ec5b0612fdc4212bc977740dcda0522fd30accd24204c:
nsenter: cannot open /hostproc/15566/ns/net: No such file or directory
```

I got this once and have not been able to reproduce it since.

Might be related to #1898, though I am running a 3.19 kernel.

--- a/net/if.go
+++ b/net/if.go
@@ -57,8 +57,7 @@ func EnsureInterfaceAndMcastRoute(ifaceName string) (*net.Interface, error) {
 		return iface, err
 	}
 	for update := range ch {
-		link, _ := netlink.LinkByIndex(update.Route.LinkIndex)
-		if link.Attrs().Name == ifaceName && update.Route.Dst.IP.Equal(dest) {
+		if update.Route.LinkIndex == iface.Index && update.Route.Dst.IP.Equal(dest) {
 			break
 		}
 	}

--- a/net/if.go
+++ b/net/if.go
@@ -52,15 +52,25 @@ func EnsureInterfaceAndMcastRoute(ifaceName string) (*net.Interface, error) {
 		return nil, err
 	}
 	dest := net.IPv4(224, 0, 0, 0)
-	// check for currently-existing route after subscribing, to avoid race
-	if CheckRouteExists(ifaceName, dest) {
-		return iface, err
+	check := func(route netlink.Route) bool {
+		return route.LinkIndex == iface.Index && route.Dst.IP.Equal(dest)
 	}
-	for update := range ch {
-		if update.Route.LinkIndex == iface.Index && update.Route.Dst.IP.Equal(dest) {
-			break
+	// check for currently-existing route after subscribing, to avoid race
+	routes, err := netlink.RouteList(nil, netlink.FAMILY_V4)
+	if err != nil {
+		return nil, err
+	}
+	for _, route := range routes {
+		if check(route) {
+			return iface, nil
 		}
 	}
+	for update := range ch {
+		if check(update.Route) {
+			return iface, nil
+		}
+	}
+	// should never get here
 	return iface, nil
 }
 

--- a/net/route.go
+++ b/net/route.go
@@ -54,14 +54,3 @@ func forEachRoute(ignoreIfaceNames map[string]struct{}, check func(name string, 
 	}
 	return nil
 }
-
-func CheckRouteExists(ifaceName string, dest net.IP) bool {
-	found := false
-	forEachRoute(map[string]struct{}{}, func(name string, route netlink.Route) error {
-		if name == ifaceName && route.Dst.IP.Equal(dest) {
-			found = true
-		}
-		return nil
-	})
-	return found
-}

--- a/net/route.go
+++ b/net/route.go
@@ -41,12 +41,14 @@ func forEachRoute(ignoreIfaceNames map[string]struct{}, check func(name string, 
 	}
 	for _, route := range routes {
 		link, err := netlink.LinkByIndex(route.LinkIndex)
-		if err == nil {
-			if _, found := ignoreIfaceNames[link.Attrs().Name]; found {
-				continue
-			}
+		if err != nil {
+			continue
 		}
-		if err := check(link.Attrs().Name, route); err != nil {
+		ifaceName := link.Attrs().Name
+		if _, found := ignoreIfaceNames[ifaceName]; found {
+			continue
+		}
+		if err := check(ifaceName, route); err != nil {
 			return err
 		}
 	}

--- a/net/route.go
+++ b/net/route.go
@@ -10,7 +10,7 @@ import (
 // A network is considered free if it does not overlap any existing
 // routes on this host. This is the same approach taken by Docker.
 func CheckNetworkFree(subnet *net.IPNet, ignoreIfaceNames map[string]struct{}) error {
-	return forEachRoute(ignoreIfaceNames, func(name string, route netlink.Route) error {
+	return forEachRoute(ignoreIfaceNames, func(route netlink.Route) error {
 		if route.Dst != nil && overlaps(route.Dst, subnet) {
 			return fmt.Errorf("Network %s overlaps with existing route %s on host.", subnet, route.Dst)
 		}
@@ -26,7 +26,7 @@ func overlaps(n1, n2 *net.IPNet) bool {
 // For a specific address, we only care if it is actually *inside* an
 // existing route, because weave-local traffic never hits IP routing.
 func CheckAddressOverlap(addr net.IP, ignoreIfaceNames map[string]struct{}) error {
-	return forEachRoute(ignoreIfaceNames, func(name string, route netlink.Route) error {
+	return forEachRoute(ignoreIfaceNames, func(route netlink.Route) error {
 		if route.Dst != nil && route.Dst.Contains(addr) {
 			return fmt.Errorf("Address %s overlaps with existing route %s on host.", addr, route.Dst)
 		}
@@ -34,21 +34,22 @@ func CheckAddressOverlap(addr net.IP, ignoreIfaceNames map[string]struct{}) erro
 	})
 }
 
-func forEachRoute(ignoreIfaceNames map[string]struct{}, check func(name string, r netlink.Route) error) error {
+func forEachRoute(ignoreIfaceNames map[string]struct{}, check func(r netlink.Route) error) error {
+	ignoreIfaceIndices := make(map[int]struct{})
+	for ifaceName := range ignoreIfaceNames {
+		if iface, err := net.InterfaceByName(ifaceName); err == nil {
+			ignoreIfaceIndices[iface.Index] = struct{}{}
+		}
+	}
 	routes, err := netlink.RouteList(nil, netlink.FAMILY_V4)
 	if err != nil {
 		return err
 	}
 	for _, route := range routes {
-		link, err := netlink.LinkByIndex(route.LinkIndex)
-		if err != nil {
+		if _, found := ignoreIfaceIndices[route.LinkIndex]; found {
 			continue
 		}
-		ifaceName := link.Attrs().Name
-		if _, found := ignoreIfaceNames[ifaceName]; found {
-			continue
-		}
-		if err := check(ifaceName, route); err != nil {
+		if err := check(route); err != nil {
 			return err
 		}
 	}

